File: src/go/renewal.go
package oreslocks

import (
	"context"
	"errors"
	"fmt"
	"time"
)

// MaxRenewalClockMS is the logical-millisecond domain shared with browser SDKs.
const MaxRenewalClockMS int64 = 9_007_199_254_740_991

// MaxRenewalTTLMS is the common TTL domain supported without duration overflow.
const MaxRenewalTTLMS int64 = 9_223_372_036_854

// RenewalPolicy controls when a held lease is renewed.
type RenewalPolicy struct {
	RenewEvery   time.Duration
	SafetyMargin time.Duration
}

// DefaultRenewalPolicy renews a 60-second lease at 20 seconds or ten seconds
// before its local deadline, whichever comes first.
func DefaultRenewalPolicy() RenewalPolicy {
	return RenewalPolicy{RenewEvery: 20 * time.Second, SafetyMargin: 10 * time.Second}
}

// RenewalLossReason is why a supervisor permanently stopped authorizing work.
type RenewalLossReason string

const (
	RenewalInvalidPolicy           RenewalLossReason = "invalid_policy"
	RenewalClockRegression         RenewalLossReason = "clock_regression"
	RenewalExpired                 RenewalLossReason = "expired"
	RenewalFailed                  RenewalLossReason = "renewal_failed"
	RenewalIdentityChanged         RenewalLossReason = "identity_changed"
	RenewalTokenChanged            RenewalLossReason = "token_changed"
	RenewalDeadlineMissing         RenewalLossReason = "deadline_missing"
	RenewalDeadlineInvalid         RenewalLossReason = "deadline_invalid"
	RenewalDeadlineRegressed       RenewalLossReason = "deadline_regressed"
	RenewalCompletionAfterDeadline RenewalLossReason = "completion_after_deadline"
	RenewalDeadlineOverflow        RenewalLossReason = "deadline_overflow"
	RenewalInvalidTTL              RenewalLossReason = "invalid_ttl"
)

// RenewalError is terminal and sticky for one supervisor.
type RenewalError struct {
	Reason  RenewalLossReason
	Message string
	Cause   error
}

func (e *RenewalError) Error() string {
	if e.Cause != nil {
		return fmt.Sprintf("%s: %s; cause: %v", e.Reason, e.Message, e.Cause)
	}
	return fmt.Sprintf("%s: %s", e.Reason, e.Message)
}

func (e *RenewalError) Unwrap() error { return e.Cause }

// RenewalDecisionKind is the pure scheduling outcome.
type RenewalDecisionKind string

const (
	RenewalWait RenewalDecisionKind = "wait"
	RenewalNow  RenewalDecisionKind = "renew_now"
	RenewalLost RenewalDecisionKind = "lost"
)

// RenewalDecision is returned by Decision.
type RenewalDecision struct {
	Kind      RenewalDecisionKind
	CheckInMS int64
	Reason    RenewalLossReason
}

// RenewalCheckpointKind is a successful checkpoint outcome.
type RenewalCheckpointKind string

const (
	RenewalCheckpointWait    RenewalCheckpointKind = "wait"
	RenewalCheckpointRenewed RenewalCheckpointKind = "renewed"
)

// RenewalCheckpoint is returned after waiting or validating a renewal.
type RenewalCheckpoint struct {
	Kind      RenewalCheckpointKind
	CheckInMS int64
}

// RenewalClock reports process-local monotonic milliseconds.
type RenewalClock interface{ NowMillis() int64 }

// RenewalClockFunc adapts a function to RenewalClock.
type RenewalClockFunc func() int64

func (clock RenewalClockFunc) NowMillis() int64 { return clock() }

// RenewalSupervisor tracks one exact Fiducia grant identity.
type RenewalSupervisor struct {
	grant           LeaseGrant
	policy          RenewalPolicy
	localDeadlineMS int64
	nextRenewalMS   int64
	lastObservedMS  int64
	loss            *RenewalError
}

// NewRenewalSupervisor starts supervision at a monotonic logical time.
func NewRenewalSupervisor(grant LeaseGrant, policy RenewalPolicy, nowMS int64) (*RenewalSupervisor, error) {
	if err := validateRenewalClock(nowMS); err != nil {
		return nil, err
	}
	if err := validateAuthorityDeadline(grant.LeaseExpiresMs); err != nil {
		return nil, err
	}
	deadline, next, err := renewalSchedule(nowMS, grant.TTLMs, policy)
	if err != nil {
		return nil, err
	}
	return &RenewalSupervisor{
		grant:           grant,
		policy:          policy,
		localDeadlineMS: deadline,
		nextRenewalMS:   next,
		lastObservedMS:  nowMS,
	}, nil
}

func (s *RenewalSupervisor) Grant() LeaseGrant      { return s.grant }
func (s *RenewalSupervisor) LocalDeadlineMS() int64 { return s.localDeadlineMS }
func (s *RenewalSupervisor) NextRenewalMS() int64   { return s.nextRenewalMS }
func (s *RenewalSupervisor) Loss() *RenewalError    { return s.loss }
func (s *RenewalSupervisor) IsLive() bool           { return s.loss == nil }

// Decision observes the monotonic clock without contacting the authority.
func (s *RenewalSupervisor) Decision(nowMS int64) RenewalDecision {
	if s.loss != nil {
		return RenewalDecision{Kind: RenewalLost, Reason: s.loss.Reason}
	}
	if err := validateRenewalClock(nowMS); err != nil {
		return s.lose(err)
	}
	if nowMS < s.lastObservedMS {
		return s.lose(&RenewalError{Reason: RenewalClockRegression, Message: fmt.Sprintf("monotonic clock regressed from %d to %d", s.lastObservedMS, nowMS)})
	}
	s.lastObservedMS = nowMS
	if nowMS >= s.localDeadlineMS {
		return s.lose(&RenewalError{Reason: RenewalExpired, Message: fmt.Sprintf("local lease deadline %d was reached at %d", s.localDeadlineMS, nowMS)})
	}
	if nowMS >= s.nextRenewalMS {
		return RenewalDecision{Kind: RenewalNow}
	}
	return RenewalDecision{Kind: RenewalWait, CheckInMS: s.nextRenewalMS - nowMS}
}

// AssertLive fails after expiry, clock regression, or any earlier renewal loss.
func (s *RenewalSupervisor) AssertLive(nowMS int64) error {
	decision := s.Decision(nowMS)
	if decision.Kind == RenewalLost {
		return s.loss
	}
	return nil
}

// Checkpoint renews only when due. Any failure permanently cancels authority.
func (s *RenewalSupervisor) Checkpoint(ctx context.Context, lease Lease, clock RenewalClock) (RenewalCheckpoint, error) {
	decision := s.Decision(clock.NowMillis())
	switch decision.Kind {
	case RenewalWait:
		return RenewalCheckpoint{Kind: RenewalCheckpointWait, CheckInMS: decision.CheckInMS}, nil
	case RenewalLost:
		return RenewalCheckpoint{}, s.loss
	}
	previous := s.grant
	renewed, err := lease.Renew(ctx, previous, time.Duration(previous.TTLMs)*time.Millisecond)
	if err != nil {
		return RenewalCheckpoint{}, s.fail(&RenewalError{Reason: RenewalFailed, Message: "lease authority did not prove continued ownership", Cause: err})
	}
	return s.AcceptRenewal(clock.NowMillis(), renewed)
}

// AcceptRenewal validates one authority success and advances the local window.
func (s *RenewalSupervisor) AcceptRenewal(completedMS int64, renewed LeaseGrant) (RenewalCheckpoint, error) {
	if s.loss != nil {
		return RenewalCheckpoint{}, s.loss
	}
	if err := validateRenewalClock(completedMS); err != nil {
		return RenewalCheckpoint{}, s.fail(err)
	}
	if completedMS < s.lastObservedMS {
		return RenewalCheckpoint{}, s.fail(&RenewalError{Reason: RenewalClockRegression, Message: fmt.Sprintf("monotonic clock regressed from %d to %d during renewal", s.lastObservedMS, completedMS)})
	}
	s.lastObservedMS = completedMS
	if completedMS >= s.localDeadlineMS {
		return RenewalCheckpoint{}, s.fail(&RenewalError{Reason: RenewalCompletionAfterDeadline, Message: fmt.Sprintf("renewal completed at %d, not before local deadline %d", completedMS, s.localDeadlineMS)})
	}
	if renewed.Key != s.grant.Key || renewed.Holder != s.grant.Holder {
		return RenewalCheckpoint{}, s.fail(&RenewalError{Reason: RenewalIdentityChanged, Message: "renewal changed the lock key or holder identity"})
	}
	if renewed.FencingToken != s.grant.FencingToken {
		return RenewalCheckpoint{}, s.fail(&RenewalError{Reason: RenewalTokenChanged, Message: fmt.Sprintf("renewal changed fencing token %d to %d", s.grant.FencingToken, renewed.FencingToken)})
	}
	if err := validateDeadlineProgress(s.grant.LeaseExpiresMs, renewed.LeaseExpiresMs); err != nil {
		return RenewalCheckpoint{}, s.fail(err)
	}
	deadline, next, err := renewalSchedule(completedMS, renewed.TTLMs, s.policy)
	if err != nil {
		return RenewalCheckpoint{}, s.fail(err)
	}
	s.grant = renewed
	s.localDeadlineMS = deadline
	s.nextRenewalMS = next
	return RenewalCheckpoint{Kind: RenewalCheckpointRenewed, CheckInMS: next - completedMS}, nil
}

func (s *RenewalSupervisor) lose(err error) RenewalDecision {
	terminal := s.fail(err)
	return RenewalDecision{Kind: RenewalLost, Reason: terminal.(*RenewalError).Reason}
}

func (s *RenewalSupervisor) fail(err error) error {
	if s.loss == nil {
		var renewal *RenewalError
		if errors.As(err, &renewal) {
			s.loss = renewal
		} else {
			s.loss = &RenewalError{Reason: RenewalFailed, Message: "renewal failed", Cause: err}
		}
	}
	return s.loss
}

func validateRenewalClock(value int64) error {
	if value < 0 || value > MaxRenewalClockMS {
		return &RenewalError{Reason: RenewalDeadlineOverflow, Message: fmt.Sprintf("logical clock must be within 0..=%d", MaxRenewalClockMS)}
	}
	return nil
}

func validateAuthorityDeadline(value int64) error {
	if value < 0 || value > MaxRenewalClockMS {
		return &RenewalError{Reason: RenewalDeadlineInvalid, Message: "authority deadline is outside the shared millisecond domain"}
	}
	return nil
}

func validateDeadlineProgress(previous, renewed int64) error {
	if err := validateAuthorityDeadline(renewed); err != nil {
		return err
	}
	if previous > 0 && renewed == 0 {
		return &RenewalError{Reason: RenewalDeadlineMissing, Message: "renewal omitted a deadline that the authority previously reported"}
	}
	if previous > 0 && renewed <= previous {
		return &RenewalError{Reason: RenewalDeadlineRegressed, Message: fmt.Sprintf("authority deadline did not advance: %d -> %d", previous, renewed)}
	}
	return nil
}

func renewalSchedule(nowMS, ttlMS int64, policy RenewalPolicy) (int64, int64, error) {
	if ttlMS <= 0 || ttlMS > MaxRenewalTTLMS {
		return 0, 0, &RenewalError{Reason: RenewalInvalidTTL, Message: fmt.Sprintf("lease TTL must be within 1..=%d ms", MaxRenewalTTLMS)}
	}
	renewEveryMS := policy.RenewEvery.Milliseconds()
	safetyMarginMS := policy.SafetyMargin.Milliseconds()
	if renewEveryMS <= 0 || safetyMarginMS <= 0 || renewEveryMS >= ttlMS || safetyMarginMS >= ttlMS {
		return 0, 0, &RenewalError{Reason: RenewalInvalidPolicy, Message: "renewal interval and safety margin must both be positive and less than the lease TTL"}
	}
	if nowMS > MaxRenewalClockMS-ttlMS || nowMS > MaxRenewalClockMS-renewEveryMS {
		return 0, 0, &RenewalError{Reason: RenewalDeadlineOverflow, Message: "renewal schedule exceeds the shared logical-clock domain"}
	}
	deadline := nowMS + ttlMS
	intervalDue := nowMS + renewEveryMS
	marginDue := deadline - safetyMarginMS
	next := intervalDue
	if marginDue < next {
		next = marginDue
	}
	if next <= nowMS {
		return 0, 0, &RenewalError{Reason: RenewalInvalidPolicy, Message: "renewal policy leaves no positive live interval"}
	}
	return deadline, next, nil
}

File: src/ts/src/renewal.ts
/**
 * Deterministic, fail-closed supervision for long-running Fiducia leases.
 *
 * This is cooperative cancellation: call `checkpoint` before each protected
 * commit. A heartbeat cannot make an unfenced external side effect safe and
 * cannot undo an effect emitted before the checkpoint.
 */

import type { Lease, LeaseGrant } from "./lease.js";

export const MAX_RENEWAL_CLOCK_MS = 9_007_199_254_740_991;
/** Largest TTL exactly representable by every runtime duration type. */
export const MAX_RENEWAL_TTL_MS = 9_223_372_036_854;

export interface RenewalPolicy {
  readonly renewEveryMs: number;
  readonly safetyMarginMs: number;
}

export const DEFAULT_RENEWAL_POLICY: RenewalPolicy = {
  renewEveryMs: 20_000,
  safetyMarginMs: 10_000,
};

export type RenewalLossReason =
  | "invalid_policy"
  | "clock_regression"
  | "expired"
  | "renewal_failed"
  | "identity_changed"
  | "token_changed"
  | "deadline_missing"
  | "deadline_invalid"
  | "deadline_regressed"
  | "completion_after_deadline"
  | "deadline_overflow"
  | "invalid_ttl";

export type RenewalDecision =
  | { readonly kind: "wait"; readonly checkInMs: number }
  | { readonly kind: "renew_now" }
  | { readonly kind: "lost"; readonly reason: RenewalLossReason };

export type RenewalCheckpoint =
  | { readonly kind: "wait"; readonly checkInMs: number }
  | { readonly kind: "renewed"; readonly checkInMs: number };

export class RenewalError extends Error {
  readonly reason: RenewalLossReason;
  override readonly cause: unknown;

  constructor(reason: RenewalLossReason, message: string, cause?: unknown) {
    super(`${reason}: ${message}`, { cause });
    this.name = "RenewalError";
    this.reason = reason;
    this.cause = cause;
  }
}

export type MonotonicClock = () => number;

/** Process-relative monotonic milliseconds for browsers and Node.js. */
export function monotonicNowMs(): number {
  return Math.trunc(performance.now());
}

export class LeaseRenewalSupervisor {
  #grant: LeaseGrant;
  readonly #policy: RenewalPolicy;
  #localDeadlineMs: number;
  #nextRenewalMs: number;
  #lastObservedMs: number;
  #loss: RenewalError | undefined;

  constructor(grant: LeaseGrant, policy: RenewalPolicy, nowMs: number) {
    validateGrant(grant);
    validateClock(nowMs);
    validateAuthorityDeadline(grant.leaseExpiresMs);
    const schedule = makeSchedule(nowMs, grant.ttlMs, policy);
    this.#grant = grant;
    this.#policy = policy;
    this.#localDeadlineMs = schedule.deadlineMs;
    this.#nextRenewalMs = schedule.nextRenewalMs;
    this.#lastObservedMs = nowMs;
  }

  get grant(): LeaseGrant {
    return this.#grant;
  }

  get localDeadlineMs(): number {
    return this.#localDeadlineMs;
  }

  get nextRenewalMs(): number {
    return this.#nextRenewalMs;
  }

  get loss(): RenewalError | undefined {
    return this.#loss;
  }

  get isLive(): boolean {
    return this.#loss === undefined;
  }

  decide(nowMs: number): RenewalDecision {
    if (this.#loss) return { kind: "lost", reason: this.#loss.reason };
    try {
      validateClock(nowMs);
    } catch (error) {
      return this.#lose(asRenewalError(error));
    }
    if (nowMs < this.#lastObservedMs) {
      return this.#lose(
        new RenewalError(
          "clock_regression",
          `monotonic clock regressed from ${this.#lastObservedMs} to ${nowMs}`,
        ),
      );
    }
    this.#lastObservedMs = nowMs;
    if (nowMs >= this.#localDeadlineMs) {
      return this.#lose(
        new RenewalError(
          "expired",
          `local lease deadline ${this.#localDeadlineMs} was reached at ${nowMs}`,
        ),
      );
    }
    if (nowMs >= this.#nextRenewalMs) return { kind: "renew_now" };
    return { kind: "wait", checkInMs: this.#nextRenewalMs - nowMs };
  }

  assertLive(nowMs: number): void {
    if (this.decide(nowMs).kind === "lost") throw this.#loss;
  }

  async checkpoint(lease: Lease, clock: MonotonicClock = monotonicNowMs): Promise<RenewalCheckpoint> {
    const decision = this.decide(clock());
    if (decision.kind === "wait") return decision;
    if (decision.kind === "lost") throw this.#loss;

    const previous = this.#grant;
    let renewed: LeaseGrant;
    try {
      renewed = await Promise.resolve().then(() => lease.renew(previous, previous.ttlMs));
    } catch (cause) {
      throw this.#fail(
        new RenewalError(
          "renewal_failed",
          "lease authority did not prove continued ownership",
          cause,
        ),
      );
    }
    return this.acceptRenewal(clock(), renewed);
  }

  acceptRenewal(completedMs: number, renewed: LeaseGrant): RenewalCheckpoint {
    if (this.#loss) throw this.#loss;
    try {
      validateClock(completedMs);
      validateGrant(renewed);
    } catch (error) {
      throw this.#fail(asRenewalError(error));
    }
    if (completedMs < this.#lastObservedMs) {
      throw this.#fail(
        new RenewalError(
          "clock_regression",
          `monotonic clock regressed from ${this.#lastObservedMs} to ${completedMs} during renewal`,
        ),
      );
    }
    this.#lastObservedMs = completedMs;
    if (completedMs >= this.#localDeadlineMs) {
      throw this.#fail(
        new RenewalError(
          "completion_after_deadline",
          `renewal completed at ${completedMs}, not before local deadline ${this.#localDeadlineMs}`,
        ),
      );
    }
    if (renewed.key !== this.#grant.key || renewed.holder !== this.#grant.holder) {
      throw this.#fail(
        new RenewalError("identity_changed", "renewal changed the lock key or holder identity"),
      );
    }
    if (renewed.fencingToken !== this.#grant.fencingToken) {
      throw this.#fail(
        new RenewalError(
          "token_changed",
          `renewal changed fencing token ${this.#grant.fencingToken} to ${renewed.fencingToken}`,
        ),
      );
    }
    try {
      validateDeadlineProgress(this.#grant.leaseExpiresMs, renewed.leaseExpiresMs);
      const schedule = makeSchedule(completedMs, renewed.ttlMs, this.#policy);
      this.#grant = renewed;
      this.#localDeadlineMs = schedule.deadlineMs;
      this.#nextRenewalMs = schedule.nextRenewalMs;
      return { kind: "renewed", checkInMs: schedule.nextRenewalMs - completedMs };
    } catch (error) {
      throw this.#fail(asRenewalError(error));
    }
  }

  #lose(error: RenewalError): RenewalDecision {
    const recorded = this.#record(error);
    return { kind: "lost", reason: recorded.reason };
  }

  #fail(error: RenewalError): RenewalError {
    return this.#record(error);
  }

  #record(error: RenewalError): RenewalError {
    this.#loss ??= error;
    return this.#loss;
  }
}

function validateGrant(grant: LeaseGrant): void {
  if (!grant || typeof grant !== "object") {
    throw new RenewalError("identity_changed", "renewal response is not a grant object");
  }
  if (typeof grant.key !== "string" || typeof grant.holder !== "string" || grant.holder.length === 0) {
    throw new RenewalError("identity_changed", "grant key and holder must be non-empty strings");
  }
  if (typeof grant.fencingToken !== "bigint" || grant.fencingToken < 0n || grant.fencingToken > 18_446_744_073_709_551_615n) {
    throw new RenewalError("token_changed", "fencing token must be an unsigned 64-bit bigint");
  }
  if (!isSafeMillisecond(grant.ttlMs) || grant.ttlMs === 0 || grant.ttlMs > MAX_RENEWAL_TTL_MS) {
    throw new RenewalError("invalid_ttl", `lease TTL must be within 1..=${MAX_RENEWAL_TTL_MS} ms`);
  }
  validateAuthorityDeadline(grant.leaseExpiresMs);
}

function validateClock(value: number): void {
  if (!isSafeMillisecond(value)) {
    throw new RenewalError(
      "deadline_overflow",
      `logical clock must be an integer within 0..=${MAX_RENEWAL_CLOCK_MS}`,
    );
  }
}

function validateAuthorityDeadline(value: number | undefined): void {
  if (value === undefined) return;
  if (!isSafeMillisecond(value) || value === 0) {
    throw new RenewalError("deadline_invalid", "authority deadline must be a positive safe integer when present");
  }
}

function validateDeadlineProgress(previous: number | undefined, renewed: number | undefined): void {
  validateAuthorityDeadline(renewed);
  if (previous !== undefined && renewed === undefined) {
    throw new RenewalError(
      "deadline_missing",
      "renewal omitted a deadline that the authority previously reported",
    );
  }
  if (previous !== undefined && renewed !== undefined && renewed <= previous) {
    throw new RenewalError(
      "deadline_regressed",
      `authority deadline did not advance: ${previous} -> ${renewed}`,
    );
  }
}

function makeSchedule(nowMs: number, ttlMs: number, policy: RenewalPolicy): { deadlineMs: number; nextRenewalMs: number } {
  validateClock(nowMs);
  if (!isSafeMillisecond(ttlMs) || ttlMs === 0 || ttlMs > MAX_RENEWAL_TTL_MS) {
    throw new RenewalError("invalid_ttl", `lease TTL must be within 1..=${MAX_RENEWAL_TTL_MS} ms`);
  }
  if (
    !isPositiveSafeInteger(policy.renewEveryMs) ||
    !isPositiveSafeInteger(policy.safetyMarginMs) ||
    policy.renewEveryMs >= ttlMs ||
    policy.safetyMarginMs >= ttlMs
  ) {
    throw new RenewalError(
      "invalid_policy",
      "renewal interval and safety margin must both be positive and less than the lease TTL",
    );
  }
  const deadlineMs = nowMs + ttlMs;
  const intervalDue = nowMs + policy.renewEveryMs;
  if (!isSafeMillisecond(deadlineMs) || !isSafeMillisecond(intervalDue)) {
    throw new RenewalError("deadline_overflow", "renewal schedule exceeds the shared logical-clock domain");
  }
  const nextRenewalMs = Math.min(intervalDue, deadlineMs - policy.safetyMarginMs);
  if (nextRenewalMs <= nowMs) {
    throw new RenewalError("invalid_policy", "renewal policy leaves no positive live interval");
  }
  return { deadlineMs, nextRenewalMs };
}

function isSafeMillisecond(value: unknown): value is number {
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 0 && value <= MAX_RENEWAL_CLOCK_MS;
}

function isPositiveSafeInteger(value: unknown): value is number {
  return isSafeMillisecond(value) && value > 0;
}

function asRenewalError(error: unknown): RenewalError {
  return error instanceof RenewalError
    ? error
    : new RenewalError("renewal_failed", "renewal validation failed", error);
}
